--- downloads/PyQt-x11-gpl-4.9.1/configure.py
+++ release/PyQt3Support-r5-pre-PyQt4.9.1-gpl/configure.py
@@ -377,6 +377,7 @@
         check_module("QtXmlPatterns", "qxmlname.h", "new QXmlName()")
         check_module("phonon", "phonon/videowidget.h",
                 "new Phonon::VideoWidget()")
+        check_module("Qt3Support", "q3painter.h", "new Q3Painter()")
         check_module("QtAssistant", "qassistantclient.h",
                 "new QAssistantClient(\"foo\")", extra_lib_dirs=ass_lib_dirs,
                 extra_libs=ass_libs)
@@ -547,6 +548,9 @@
         if "QAxContainer" in pyqt_modules:
             generate_code("QAxContainer")
 
+        if "Qt3Support" in pyqt_modules:
+            generate_code("Qt3Support")
+
         # Generate the composite module.
         qtmod_sipdir = os.path.join("sip", "Qt")
         mk_clean_dir(qtmod_sipdir)
@@ -1524,6 +1528,11 @@
     # yet so we make sure we avoid the problem by always releasing the GIL.
     qt_sip_flags.append("-g")
 
+    # If the module is disabled, disable also the feature, so that qt3 support
+    # functions in other modules are compiled out.
+    if "Qt3Support" not in pyqt_modules:
+        qt_sip_flags.append("-x")
+        qt_sip_flags.append("Qt_QT3SUPPORT")
 
 def needed_qt_libs(mname, qt_libs):
     """Add any additional Qt libraries needed by a module to an existing list.
@@ -1555,6 +1564,7 @@
         "phonon": ["QtGui"],
         "QtAssistant": ["QtNetwork", "QtGui"],
         "QtDesigner": ["QtGui"],
+        "Qt3Support": ["QtNetwork", "QtSql", "QtXml"],
         "QAxContainer": ["QtGui"]
     }
 
@@ -1726,6 +1736,10 @@
 
     add_makefile_extras(makefile, extra_include_dirs, extra_lib_dirs, extra_libs)
 
+    if "Qt3Support" in pyqt_modules:
+        sipconfig.inform("QT3_SUPPORT defined for the %s module." % mname)
+        makefile.extra_defines.append("QT3_SUPPORT")
+
     makefile.generate()
 
 
